### ISSUE TITLE
File uploads via OC web interface causing a segmentation fault on Apache/2.4.10 w/ mod_fcgid + suEXEC, PHP7

### ISSUE THREAD

**Issue by HackintoshHD** [owncloud/core#22036]
**Symptomes:** File uploads of files with a size above **~500 MB** are causing a segmentation fault and subsequently fail on Apache/2.4.10 with mod_fcgid and suEXEC.

_**Caveat 1:** I am quite aware of the fact that segmentation faults are often caused by triggering an existing bug in either PHP or Apache. Nonetheless, I assume it's in the best interest of the ownCloud community to track down such a reproducible bug, as mod_fcgid and suEXEC are still essential for shared hosting._

_**Caveat 2:** I am also aware that the ownCloud Desktop Sync Client is uploading files as 10 MB chunks and therefore would not be affected by the problem. However, I still would like to resolve it._

**Already existing, possibly related reports:**
- #10675 by @Pablo1978, accompagnied by his/her German forum posts ([1](https://forum.owncloud.org/viewtopic.php?f=21&t=23093),[2](https://forum.owncloud.org/viewtopic.php?f=29&t=23312))

**System configuration:**
- **Debian Jessie** 8.2.0 (x64) on an Intel J1900 with 16 GB RAM
- **Apache**/2.4.10 with **mod_fcgid** (formerly known as 'FastCGI') and **suEXEC-custom** (its docroot configured to `/srv/www`),
- **PHP 7**.0.x from the [Dotdeb repository](https://www.dotdeb.org/2016/01/08/php-7-0-2-for-jessie/),
- **Redis**-Cache compiled and activated, basically following the steps of @enoch85 's script, modifying the module's activation in fcgid, of course.

The whole thing runs pretty darn fast and without any serious glitches - except for the file upload problem for ~500 MB+ files described here. The server is already configured in accordance with the ownCloud Server Administration manual, chapter [Uploading big files > 512MB (as set by default)](https://doc.owncloud.org/server/7.0/admin_manual/configuration/big_file_upload_configuration.html), of course.

**Details:**
- Files with a size _below_ the ~500 MB limit are uploaded fine (they're written into `/tmp` as e.g.<br>
`-rw-------  1 www-data www-data  529935856 Jan 30 19:02 fcgid.tmp.DcRrkz` and then moved over to ownCloud's data directory).<br><br>
- Files with a size _above_ the ~500 MB limit fail to upload via ownCloud's web interface (they're entirely written into `/tmp`, as well, but obviously are not copied or moved over to the ownCloud data directory).<br><br>
- The logged error message on such a failed upload in `owncloud.log`:<br>
`{"reqId":"P0110iOHs\/RqRpRbscbE","remoteAddr":"192.168.23.33","app":"files","message":"Upload error 3 - the uploaded file was only partially uploaded","level":3,"time":"January 30, 2016 19:18:48","method":"POST","url":"\/index.php\/apps\/files\/ajax\/upload.php"} `<br><br>
- The logged error message on such a failed upload in `/var/log/apache2/error.log`:<br>
`[core:notice] [pid 780:tid 139929551775616] AH00052: child pid 809 exit signal Segmentation fault (11)`<br><br>

**Further options for troubleshooting:**

I've tried to install and configure **gdb** to receive dumps of the segmentation faults for further analysis, mainly following the instructions [here](http://sysadmin.carlusgg.com/?p=197) and [here](http://www.exanto.de/apache-segfault-unter-debian-wheezy-debuggen.html) (in German). This unfortunately has not been successful yet, as no dump files are created in my CoreDumpDirectory `/tmp/apache2-coredumps`. But any hints on how to set up gdb and Apache 2.4 for getting dumps are highly appreciated …

**Comment by karlitschek** [comment]
So ownCloud is written in PHP which runs on top of the php/apache binaries. It is not possible for a script to provoke a segmentation fault in php itself.
This is most likely a problem with your php, webserver, distribution, packages, hardware or something else in the environment. Can't be a bug in owncloud itself. Please look at owncloud.org/support for places to get support.

**Comment by HackintoshHD** [comment]
<blockquote>So ownCloud is written in PHP which runs on top of the php/apache binaries. It is not possible for a script to provoke a segmentation fault in php itself.
This is most likely a problem with your php, webserver, distribution, packages, hardware or something else in the environment. […] Please look at owncloud.org/support for places to get support.</blockquote>

@karlitschek Instead of closing my reported issue right away, why can't you make it at least an issue for ownCloud's documentation, please, e.g. [here](https://doc.owncloud.org/server/8.2/admin_manual/configuration_files/big_file_upload_configuration.html) or [here](https://doc.owncloud.org/server/9.0/admin_manual/configuration_files/big_file_upload_configuration.html)?

I don't expect anyone from the ownCloud Community to fix bugs in other components like Apache or PHP, but I was at least hoping for an exchange of expertise on how to track down the underlying bug in (probably) either Apache or PHP, as it clearly _is_ affecting the operation of ownCloud in shared hosting environments …

<blockquote>Can't be a bug in owncloud itself.</blockquote>

Sure it's not a bug in ownCloud itself, that's what I already state above: It's most likely a bug in either PHP(7) or Apache. But - _and that's my point!_ - **it _is_ affecting the project**, as uploads above 500 MB will fail over the web interface, as soon as you choose to run ownCloud on a server that is properly, securely separating instances by employing different users like _mod_fcgid_ and _suEXEC_ do.

**Comment by kscholz** [comment]
I can only agree with HackintoshHD and can confirm there is a Bug wherever it may lay.
And the point is: It is affecting owncloud, cause Uploads over HTTP will fail.
So maybe a litte "Hey guys, there is a bug. It is not our fault, but you can do this and that to come round it" would be really nice. The discussion can be made in the forums of course!

As far as i can say, the Problem is located somewhere at php_fcgi.
I have the same issue on my debian 7 box with apache 2.4 and old php5 running with suexec in secure fcgi.
The File Upload works till something between 490 and 500 MB.
Files bigger will fully upload to /tmp directory but will not be copied to cloud data dir, and will not be deleted from /tmp! So you probably can kill a server by uploading a high amount of huge files wich will make the disk full.
As soon as the file upload is complete there is a segfault.
It looks like there is some sort of timeout or buffer overrun?
I tried several config options in php and fcgi but without success.
Everything i could find with buffers and timeouts. But nothing changed.

Even if we can't fix the bug.
How is this idea:

The upload is done as filestream, is the stream to big it fails.
Would it be possible to think about splitting the file into chunks and upload small chunks as stream like the desktop clients do?
Would not only "fix" this bug, it would be great for users with slow and unstable internet connections too, cause if uploading is interrupted you could start again and your progress wouldn't be gone.

So maybe this "Bug Report" could turn into a great "Feature Request"!

**Comment by HackintoshHD** [comment]
Thanks for your feedback, @kscholz - good to know I'm not the only one experiencing this (and that I can stop thinking about downgrading PHP to a pre-7-version).

Until now, I honestly hadn't thought of the problem as a security risk, but the point you raise is indeed valid:
<blockquote>Files bigger will fully upload to /tmp directory but will not be copied to cloud data dir, and will not be deleted from /tmp! So you probably can kill a server by uploading a high amount of huge files wich will make the disk full.</blockquote>

Indeed. Same experience here. Let's simply consider these facts:
- ownCloud's **default file size** limit for uploads is **512 MB**.
- With the potential Apache/2.4-/ PHP-bug, any upload of a file size between ~500 MB and 512 MB will fail (it will be completely uploaded to `/tmp`, but will remain there as a temporary file).
- As such a failed upload will not be copied over to ownCloud's data directory, I assume it will **_not_** be counted for the user's quota (feel free to correct me, if I'm wrong).

**Bottom line:**
**Even if you _have_ set quotas**, each single one of your ownCloud instance's users can easily perform a denial-of-service attack against your ownCloud instance by continuously uploading files bigger than 500 MB but smaller than 512 MB.

**Comment by kscholz** [comment]
@HackintoshHD: To your bottom line:
Users **will do** a denial-of-service attack. Perhaps not intentionally, but just by uploading big files.
After the fgci segfaults the upload simply starts again... and again... and again...
So if your Users want to upload... They keep trying... and trying... and trying...
Oh poor storage space, you'll be wasted... ;-)

On further investigation of this problem i think i got it.
It looks like it's all related to this bug in mod_fcgid from apache: [Bug #51747](https://bz.apache.org/bugzilla/show_bug.cgi?id=51747)
To confirm it i tried to change "FcgidMaxRequestInMem" in my vhost config of owncloud to a higher value than the default of 65536 bytes. With 131072 bytes nothing changed. But... with 2147483648 bytes (2 GB...) uploading of big files "simply works".
Well, it's not a bugfix... but you can do a work-a-round of the bug by wasting some of your servers memory each time a big file upload finishes. 

As you can see in the Bug Report from mod_fcgid the bug was first reported in 2011 and has been fixed twice but it looks like it's not merged in the stable releases. Oh what a shame.
Let's hope it will merged someday... ;-)

I hope this helps you @HackintoshHD and anyone who might fall over this issue.
Reports if this work-a-round works for you too would be great.

**Comment by HackintoshHD** [comment]
@kscholz Great find, thank you so much! - I've added **FcgidMaxRequestInMem** with a value of exactly 2&nbsp;Gigabytes to my virtual host configuration file and can **confirm** your workaround is working like a charm.

I've tested the scenario with files of a size around 8&nbsp;Gigabytes, so roughly that of a DVD-9, and they're now all uploaded flawlessly over ownCloud's web interface:

![oc_8 2 2_fcgid_suexec_upload](https://cloud.githubusercontent.com/assets/2894628/13076304/bd6763e6-d4b1-11e5-8c04-b11116e3f296.png)

So it must indeed be **[mod_fcgi bug #51747](https://bz.apache.org/bugzilla/show_bug.cgi?id=51747)**. The [patch](https://bz.apache.org/bugzilla/show_bug.cgi?id=51747#c6) posted by Jan Stürtz (@stuertz on Github?) for Apache/2.3.9 mod_fcgi over at the Apache Software Foundation's Bugzilla was possibly not integrated into the Apache/2.4 mod_fcgi sources. Maybe we can advocate a bugfix over there …

**Comment by HackintoshHD** [comment]
**Just for future reference,** in case anybody with a server using Apache/2.4 mod_fcgid and suEXEC would like to reproduce the problem: Here's my [Let's-Encrypt-ified](https://letsencrypt.org/) virtual host's configuration file.

**Disclaimer:** Please note that the values below are still set ridiculously high for the purpose of debugging to exclude any possibility of a misconfiguration. They're certainly _**<ins>not</ins>**_ meant for use in a production environment.
```
<virtualhost *:80>

	include /etc/apache2/include/subdomain.my-domain.toplevel

	loglevel    info
	errorlog    /var/log/apache2/owncloud-error.log
	customlog   /var/log/apache2/owncloud-access.log combined

        BusyTimeout 18000
        IdleTimeout 18000
        IPCConnectTimeout 18000   
        IPCCommTimeout 18000
        DefaultMaxClassProcessCount 8
        ProcessLifeTime 18000
        FcgidBusyTimeout 18000
        FcgidIOTimeout 18000
        FcgidMaxRequestInMem 2147483648
        FcgidMaxRequestLen 8543666176

</virtualhost>

<ifmodule mod_ssl.c>
	<virtualhost *:443>

        include /etc/apache2/include/subdomain.my-domain.toplevel

        include /etc/apache2/include/common-ssl
		SSLCertificateFile        /etc/letsencrypt/live/subdomain.my-domain.toplevel/fullchain.pem
		SSLCertificateKeyFile     /etc/letsencrypt/live/subdomain.my-domain.toplevel/privkey.pem

		loglevel    info
		errorlog    /var/log/apache2/owncloud-ssl-error.log
		customlog   /var/log/apache2/owncloud-ssl-access.log combined

	        BusyTimeout 18000
	        IdleTimeout 18000
        	IPCConnectTimeout 18000   
        	IPCCommTimeout 18000
        	DefaultMaxClassProcessCount 8
        	ProcessLifeTime 18000
        	FcgidBusyTimeout 18000
        	FcgidIOTimeout 18000
        	FcgidMaxRequestInMem 2147483648
        	FcgidMaxRequestLen 8543666176

	</virtualhost>
</ifmodule>
```
